Select knob key/value ordering

psimyn opened the pull request:
continued from https://github.com/storybooks/storybook/pull/1449

fixes #799

Issue:

Using knobs is currently confusing. 

Example, for complex strings and null values

```
select('My Select', {
  none: null, // renders empty item
  [`Computed key`]: 'computed', 
  'Some kind of long sentence. This is ok, mainly just looks odd as a key': 'sentence',
})
```


## What I did

Use value as value
Still uses key for the key property on each `<option>`

## How to test

```
select('My Select', {
  Empty: null,
  'First item': 'The first item selected',
  second: `The item with key 'second' is selected`
})
```



Is this testable with jest or storyshots?

Does this need a new example in the kitchen sink apps?

Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/storybooks/storybook/1745)
<!-- Reviewable:end -->

danielduan commented:
Think this should go out as a minor version instead of a patch.

Is this documented anywhere and also is there an example somewhere in the cra-kitchen-sink for this?

psimyn commented:
@danielduan I've updated addons README (and rebased on latest `release/3.3`)

Not sure about adding to kitchen-sink examples, since for users it is identical to existing `select`. What do you think?

danielduan commented on addons/knobs/README.md:
```diff
@@ -217,6 +217,28 @@ const value = select(label, options, defaultValue);
+};
+const defaultValue = 'Red';
+
+const value = select(label, options, defaultValue);
```
effectively, this should be `selectV2()` with this PR change?

I don't think this alleviates the breaking change, it only defers the problem to a later date.

ndelangen replied:
yes, seems that by doing that we could release this early and deprecate the old method, and then remove the old implementation at the next major.

psimyn replied:
yep, this means the breaking change in v4 will be renaming `selectV2` to `select`

this was meant to be `selectV2`, I've updated it 👍

ndelangen commented:
@psimyn The addon has changed somewhat since this PR was rebased, mainly support for Vue and Angular was introduced, and thus a lot of code was moved.

Do you think you could fix the merge conflicts? Let me know if you need help?

psimyn committed:
key and value for select knob match key/value of options

fixes #799
add selectV2 for backwards compatibility

psimyn commented:
@ndelangen I've updated this with latest master, and added a test for the SelectType component.

Let me know if anything else is needed - the existing example is still valid (since it uses the Array of options).

apologies for the massive delay!

ndelangen committed:
Merge branch 'master' into select-knob-ordering

ndelangen committed:
Update README.md

Hypnosphi commented:
Should we add a deprecation notice for old `select` usages?

ndelangen commented:
@psimyn Thanks for the hard work! Really appreciate the time and effort you've put in!

This is a good step overall to improve the knobs API!

I feel it's merge-able without the deprecation warning on `select`, but if you feel like adding it to this PR that's great!

Merge when ready IMHO

psimyn commented:
thanks @ndelangen & @Hypnosphi 

I'll add deprecate, but could use some help with wording. Currently have:

select: 'in v4 the options keys/values are being swapped'
selectV2: 'in v4 this will be renamed to select'

ndelangen committed:
Merge branch 'master' into select-knob-ordering

ndelangen commented:
@shilman You wanna pitch on on the deprecation messages?

kevinSuttle commented:
So, how does one use these to set booleans on a component? 

Example: button variants: https://www.styled-components.com/docs/basics#adapting-based-on-props

```js
const buttonVariants = {
  primary: "primary",
  secondary: "secondary",
  negative: "negative",
  alternate: "alternate"
};

stories
  .add("Button with variants", () => (
    <Button
      onClick={action("clicked")}
      disabled={boolean("Disabled", false)}
      variants={select("Variants", buttonVariants, buttonVariants.primary)}
    >
      {text("Label", "Find a mentor from this school")}
    </Button>
  ))
```

psimyn committed:
Merge branch 'master' into select-knob-ordering

psimyn committed:
Merge branch 'master' of github.com:storybooks/storybook into select-knob-ordering

psimyn committed:
Merge branch 'select-knob-ordering' of github.com:storybooks/storybook into select-knob-ordering

psimyn committed:
add deprecation messages for select and selectV2

Hypnosphi commented:
> select: 'in v4 the options keys/values are being swapped'
> selectV2: 'in v4 this will be renamed to select'

The problem with this is that currently there is no way to avoid getting warnings. Maybe we should add the latter warning only in `4.0`, and remove `selectV2` only in `5.0`

Hypnosphi commented:
@kevinSuttle in your case, keys and values are the same string, so this PR shouldn't change anything for you.

Btw, you can use an array for that:
```
const buttonVariants = [
  "primary",
  "secondary",
  "negative",
  "alternate"
];
```

psimyn committed:
don't deprecate selectV2 until v5

psimyn committed:
Merge branch 'master' into select-knob-ordering

psimyn commented:
thanks @Hypnosphi - I didn't like always showing deprecation message. I've removed the v2 one

Hypnosphi commented on addons/knobs/src/base.js:
```diff
@@ -46,7 +47,14 @@ export function object(name, value) {
+    manager.knob(name, { type: 'select', options, value }),
+    'in v4 keys/values of the options argument are reversed'
+  );
+}
```
For some reason, it throws in unit tests:
```
TypeError: Object prototype may only be an Object or null: apple
        at Function.setPrototypeOf (<anonymous>)

      74 | 
      75 | function select(name, options, value) {
    > 76 |   return (0, _utilDeprecate2.default)(manager.knob(name, { type: 'select', options: options, value: value }), 'in v4 keys/values of the options argument are reversed');
      77 | }
      78 | 
      79 | function selectV2(name, options, value) {
      
      at select (addons/knobs/dist/base.js:76:38)
```
https://circleci.com/gh/storybooks/storybook/33655?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

psimyn replied:
I was doing it wrong :)

fixed this up and added selectV2 example

Hypnosphi reviewed (changes requested):
Please add some `selectV2` usages to `examles/official-storybook`

psimyn committed:
Merge branch 'master' of github.com:storybooks/storybook into select-knob-ordering

psimyn committed:
Merge branch 'master' of github.com:storybooks/storybook into select-knob-ordering

psimyn committed:
fix deprecate message, add example for select v2

psimyn committed:
Merge branch 'select-knob-ordering' of github.com:storybooks/storybook into select-knob-ordering

psimyn committed:
fix example

psimyn committed:
update snapshot with text from selectV2 example

ndelangen committed:
Merge branch 'master' into select-knob-ordering

ndelangen commented:
Perfect @psimyn This is good to go IMHO 👍

ndelangen commented:
Is there something we're doing wrong that is causing the deprecation message to appear multiple times?

<img width="707" alt="screen shot 2018-01-31 at 22 06 18" src="https://user-images.githubusercontent.com/3070389/35647499-0f0d93be-06d3-11e8-95ef-1ed4e2b5a407.png">

Hypnosphi commented on addons/knobs/src/base.js:
```diff
@@ -46,7 +47,14 @@ export function object(name, value) {
+    manager.knob.bind(manager),
+    'in v4 keys/values of the options argument are reversed'
+  )(name, { type: 'select', options, value });
+}
```
This should probably be rewritten as follows:

```
export const select = deprecate(
  (name, options, value) => manager.knob(name, { type: 'select', options, value }),
  'in v4 keys/values of the options argument are reversed'
);
```

See https://github.com/storybooks/storybook/pull/1745#issuecomment-362070645

psimyn replied:
thanks! I was missing the arrow function before and passing the result of `manager.knob` 🤦

psimyn committed:
deprecate properly

psimyn committed:
Merge branch 'select-knob-ordering' of github.com:storybooks/storybook into select-knob-ordering

psimyn commented:
thanks @ndelangen & @Hypnosphi 

multiple warnings was due to me deprecating incorrectly. Fixed up now 👍
